kin-devplatform-sdk/src/main/java/kin/devplatform/KinEcosystemInitiator.java

```diff
@@ -0,0 +1,265 @@
+import kin.devplatform.accountmanager.AccountManager.AccountState;
+import kin.devplatform.accountmanager.AccountManagerImpl;
+import kin.devplatform.accountmanager.AccountManagerLocal;
+import kin.devplatform.base.ObservableData;
+import kin.devplatform.base.Observer;
+import kin.devplatform.bi.EventLogger;
+import kin.devplatform.bi.EventLoggerImpl;
+import kin.devplatform.bi.events.GeneralEcosystemSdkError;
+import kin.devplatform.bi.events.KinSdkInitiated;
+import kin.devplatform.core.util.DeviceUtils;
+import kin.devplatform.core.util.ExecutorsUtil;
+import kin.devplatform.data.auth.AuthLocalData;
+import kin.devplatform.data.auth.AuthRemoteData;
+import kin.devplatform.data.auth.AuthRepository;
+import kin.devplatform.data.blockchain.BlockchainSourceImpl;
+import kin.devplatform.data.blockchain.BlockchainSourceLocal;
+import kin.devplatform.data.offer.OfferRemoteData;
+import kin.devplatform.data.offer.OfferRepository;
+import kin.devplatform.data.order.OrderLocalData;
+import kin.devplatform.data.order.OrderRemoteData;
+import kin.devplatform.data.order.OrderRepository;
+import kin.devplatform.exception.BlockchainException;
+import kin.devplatform.exception.KinEcosystemException;
+import kin.devplatform.network.model.AuthToken;
+import kin.devplatform.network.model.SignInData;
+import kin.devplatform.util.ErrorUtil;
+
+public final class KinEcosystemInitiator {
+
+	private static final String KIN_ECOSYSTEM_STORE_PREFIX_KEY = "kinecosystem_store";
+	private static final int ACCOUNT_CREATION_TIME_OUT_MILLIS = 30000;
+	private static KinEcosystemInitiator instance;
+
+	private final ExecutorsUtil executorsUtil;
+	private volatile boolean isInitialized = false;
+	private volatile boolean isLoggedIn = false;
+
+	public static KinEcosystemInitiator getInstance() {
+		if (instance == null) {
+			synchronized (KinEcosystemInitiator.class) {
+				if (instance == null) {
+					instance = new KinEcosystemInitiator();
+				}
+			}
+		}
+		return instance;
+	}
+
+	private KinEcosystemInitiator() {
+		this.executorsUtil = new ExecutorsUtil();
+	}
+
+	/**
+	 * Uses for internal initialization after process restart
+	 */
+	public void internalInit(Context context) {
+		try {
+			init(context, null);
+		} catch (BlockchainException e) {
+			EventLoggerImpl.getInstance().send(GeneralEcosystemSdkError.create(
+				ErrorUtil.getPrintableStackTrace(e), String.valueOf(e.getCode()),
+				"KinEcosystemInitiator internalInit failed"
+			));
+			e.printStackTrace();
+		}
+	}
+
+	/**
+	 * Uses for external (public API) initialization and jwt login.
+	 */
+	public void externalInit(Context context, KinEnvironment environment, @NonNull SignInData signInData,
+		final KinCallback<Void> loginCallback) {
+		if (isInitialized && isLoggedIn) {
+			fireStartCompleted(loginCallback);
+			return;
+		}
+
+		try {
+			init(context, environment);
+		} catch (BlockchainException e) {
+			fireStartError(e, loginCallback);
+		}
+		login(signInData, loginCallback);
```

Comment:
Are you sure you want to login anyway? 
For example if the try catch will go to the catch then after fireStartError the login will be invoked

Reply:
no I don't, good catch